mongorestore/mongorestore_test.go

```diff
@@ -3664,6 +3668,90 @@ func testDumpAndRestoreAllDBsIgnoresSomeConfigCollections(t *testing.T) {
 	)
 }
 
+func TestIgnoreMongoDBInternal(t *testing.T) {
+	testtype.SkipUnlessTestType(t, testtype.IntegrationTestType)
+
+	sessionProvider, _, err := testutil.GetBareSessionProvider()
+	require.NoError(t, err)
+
+	if ok, _ := sessionProvider.IsReplicaSet(); !ok {
+		t.Skip("replica set required")
+	}
+
+	ctx := t.Context()
+
+	dbName := util.MongoDBInternalDBPrefix + t.Name()
+
+	client, err := testutil.GetBareSession()
+	require.NoError(t, err, "must connect to server")
+
+	internalColl := client.Database(dbName).Collection(t.Name())
+
+	_, err = internalColl.InsertOne(ctx, bson.D{{"_id", 1}})
+	require.NoError(t, err, "must write to the internal DB")
+
+	_, err = client.Database(t.Name()).Collection(t.Name()).InsertOne(ctx, bson.D{})
+	require.NoError(t, err, "must write to the user DB")
+
+	// Send repeated updates to the internal DB in a background thread.
+	writesCtx, writesCancel := context.WithCancelCause(ctx)
+	updatesDone := make(chan struct{})
+	go func() {
+		defer close(updatesDone)
+
+		for writesCtx.Err() == nil {
+			_, err := internalColl.InsertOne(
+				writesCtx,
+				bson.D{},
+			)
+
+			if !errors.Is(err, context.Canceled) {
+				require.NoError(t, err, "must write to the internal DB")
+			}
+		}
+
+		t.Logf("Updates canceled: %v", context.Cause(writesCtx))
+	}()
+
+	withArchiveMongodump(
+		t,
+		func(archive string) {
+			writesCancel(fmt.Errorf("archive is finished"))
+			<-updatesDone
```

Comment:
Should this wait a little bit to ensure there actually _were_ writes?

Reply:
There’s a wait built in because withArchiveMongodump runs mongodump before it calls the callback:
- “background” thread starts inserting
- mongodump creates archive
- callback is invoked, which cancels the background thread’s context

This channel is just here to ensure that no writes happen during the restore.